First stab at updating token

lwasser opened the pull request:
test pr

lwasser committed:
add: summaryathlete to exchange code for token

lwasser committed:
fix: typing and update changelog

lwasser committed:
feat: add tests for new features

lwasser committed:
Fix: update tests

Co-authored-by: Jonatan Samoocha <jsamoocha@yahoo.com>

lwasser committed:
fix: refactor based on review

lwasser committed:
fix: update tests

lwasser committed:
Update src/stravalib/protocol.py

lwasser committed:
fix: typing corrections

lwasser committed:
feat(tests): add tests for client exchange as well

lwasser committed:
fix: test doctstrings

lwasser committed:
Fix: athlete could return non if strava stops returning that element

Co-authored-by: Jonatan Samoocha <jsamoocha@yahoo.com>

lwasser committed:
fix: make code more explicit

lwasser committed:
fix: make return items more explicit

lwasser committed:
Merge branch 'main' into token-exch

jsamoocha committed:
Fixes remaining typing issue

jsamoocha committed:
Minor typo in comment

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.6.9 → v0.8.6](https://github.com/astral-sh/ruff-pre-commit/compare/v0.6.9...v0.8.6)
- [github.com/adamchainz/blacken-docs: 1.19.0 → 1.19.1](https://github.com/adamchainz/blacken-docs/compare/1.19.0...1.19.1)

lwasser committed:
First stab at updating token

lwasser commented on src/stravalib/client.py:
```diff
@@ -67,6 +67,8 @@ def __init__(
         rate_limit_requests: bool = True,
         rate_limiter: limiter.RateLimiter | None = None,
         requests_session: Session | None = None,
+        token_expires: int | None = None,
```
token expires and refresh values are populated by the user when they instantiate the client object

we could just use logic that if self. token_refresh exists, then try to autorefresh?

jsamoocha replied:
Maybe it's easier to store the refresh token itself as a client property, just like the access token? And use a (private) helper method to determine when the refresh token is expired?

lwasser replied:
Sure. I can add that as well! I'll work on this a bit more over the next week. Open to any and all feedback, ideas, and improvements! Especially given you know more about the use case of an app where you are retrieving data from multiple users!

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -8,11 +8,14 @@
 from urllib.parse import urlencode, urljoin, urlunsplit
 
 import requests
+from dotenv import load_dotenv
```
in case the user wants to use a dot file. this would be a new dependency however . we could make it optional? or we could just stick to os.get_environ

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -32,6 +35,35 @@
 RequestMethod = Literal["GET", "POST", "PUT", "DELETE"]
 
 
+def refresh_expired_token(func):
```
this is the old decorator that I started with. it's a different approach

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +132,65 @@ def __init__(
+            True if token has expired, otherwise returns false
+        """
+        # Right now the issue is the self.token_expires is not being populated
+        if verbose:
```
this print statement is in the wrong place to actually work.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +132,65 @@ def __init__(
+            # Checks users environment for CLIENT_ID and CLIENT_SECRET
+            client_id = os.environ["CLIENT_ID"]
+            client_secret = os.environ["CLIENT_SECRET"]
+        # "Fail" quietly if things are setup properly
```
```suggestion
        # "Fail" quietly if things are not setup properly
```

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -193,17 +299,23 @@ def exchange_code_for_token(
                 "grant_type": "authorization_code",
             },
             method="POST",
+            refresh_expired_token=False,
```
avoids an infinite loop. i haven't triggered one of these in a long time until I worked on this pr 😆

lwasser committed:
Fix: Changelog and readme + a few doc fixes

lwasser committed:
fix: remove old decorator attempt

lwasser committed:
fix: use logging instead of print statements

lwasser committed:
typing problemzzz

lwasser committed:
Fix: a few more mypy issues

lwasser committed:
fix: rename var refresh-token

lwasser committed:
fix: cleanup following @jsamoocha review

lwasser committed:
fix: more cleanup

lwasser committed:
fix: update docs so information is in one place

lwasser committed:
fix: changelog

lwasser committed:
add: tests for token_expired helper

lwasser committed:
add: more tests for token refresh

lwasser committed:
add: pytest-xdist for parallel test runs

lwasser committed:
add: todos for test coverage

lwasser committed:
fix: mypy is finally happygit add src/stravalib/protocol.py

lwasser committed:
fix: add tests

Apply suggestions from code review

lwasser committed:
fix: test cleanup and reorg

lwasser committed:
fix: make sure setter and getter work as expected

lwasser committed:
Apply suggestions from code review

lwasser committed:
Update changelog.md

lwasser committed:
feat: Return optional SummaryAthlete via exchange code for token with user request

lwasser committed:
Merge branch 'main' into fix-dev-docs

lwasser committed:
doc: update development docs so information is in one place (no duplication)

lwasser committed:
fix: move credential check to init

lwasser committed:
fix: cleanup tests

fix: ensure envt mocked in some tests

fix: test cleanup

Update noxfile.py

Update src/stravalib/protocol.py

lwasser committed:
Update src/stravalib/protocol.py

Co-authored-by: Jonatan Samoocha <jsamoocha@yahoo.com>
